### ISSUE TITLE
FURB103 fix skips file truncation when the data to write has the wrong type

### ISSUE THREAD

**Issue by dscorbett** [astral-sh/ruff#26920]
### Summary

The fix for [`write-whole-file` (FURB103)](https://docs.astral.sh/ruff/rules/write-whole-file/) can change the program’s behavior. `open` truncates the file. If the `write` call fails because the argument is of the wrong type, nothing is written, but the file is still already truncated. The fixed version checks the type before opening the file, so if the type is wrong, the file is not truncated. The fix should be marked unsafe except when Ruff is sure the argument is of the right type for the mode. [Example](https://play.ruff.rs/6150be59-f349-4969-b0ea-c0c0dca636d5):
```console
$ cat >furb103.py <<'# EOF'
import pathlib
pathlib.Path("furb103.txt").write_text("original contents")
try:
    with open("furb103.txt", "w") as f:
        f.write(b"\103")
except TypeError as e:
    print(e)
contents = pathlib.Path("furb103.txt").read_text(encoding="utf-8")
print(f"{contents=}")
# EOF

$ python furb103.py
write() argument must be str, not bytes
contents=''

$ ruff --isolated check --preview --select FURB103 furb103.py --fix
Found 1 error (1 fixed, 0 remaining).

$ cat furb103.py
import pathlib
pathlib.Path("furb103.txt").write_text("original contents")
try:
    pathlib.Path("furb103.txt").write_text(b"\103")
except TypeError as e:
    print(e)
contents = pathlib.Path("furb103.txt").read_text(encoding="utf-8")
print(f"{contents=}")

$ python furb103.py
data must be str, not bytes
contents='original contents'
```

### Version

ruff 0.15.22 (0177a7e0d 2026-07-16)

**Comment by STiFLeR7** [comment]
Note per the [AI policy](https://github.com/astral-sh/.github/blob/main/AI_POLICY.md): the analysis below was produced with AI assistance (autonomous investigation of the current `main` source), disclosed as such rather than submitted as a PR, since I couldn't put a human in the loop who could vouch for the reasoning end-to-end.

Confirmed against current `main` (`crates/ruff_linter/src/rules/refurb/rules/write_whole_file.rs`, `generate_fix`): the fix's `Applicability` is currently computed purely from comment placement:

```rust
let applicability = if checker.comment_ranges().intersects(with_stmt.range()) {
    Applicability::Unsafe
} else {
    Applicability::Safe
};
```

It never inspects the `write()` argument's type, so it's `Safe` even when the argument obviously doesn't match the declared open mode (e.g. `open(..., "w")` + `f.write(b"...")`), reproducing exactly the truncation-vs-no-truncation behavior change from the issue.

`ruff_linter` already has a helper for exactly this kind of decision — `ResolvedPythonType`/`PythonType` in `ruff_python_semantic::analyze::type_inference`, used the same way by `FURB161` (`bit_count.rs`) and `FURB164` (`unnecessary_from_float.rs`) to downgrade a fix to `Unsafe` unless the argument's type is statically known to be compatible. `FileOpen::mode` (in `refurb/helpers.rs`) already distinguishes `WriteText`/`WriteBytes`, so the fix could gate applicability on:

```rust
let type_matches_mode = matches!(
    (open.mode, ResolvedPythonType::from(content)),
    (OpenMode::WriteText, ResolvedPythonType::Atom(PythonType::String))
        | (OpenMode::WriteBytes, ResolvedPythonType::Atom(PythonType::Bytes))
);
```

combined with the existing comment check, and only mark `Safe` when both hold. This requires threading the `write()` call's argument expression (already available in `WriteMatcher::visit_expr` as `content`) into `generate_fix`, which doesn't currently receive it.

I verified this approach compiles, passes `cargo clippy -D warnings`, and produces the expected snapshot diff against the existing FURB103 test fixtures (literal `str`/`bytes` arguments stay `Safe`; every case with an unresolvable or mismatched argument type — including the `foobar`/dynamic-content cases already in `FURB103_0.py` — correctly flips to `Unsafe` with a "may change runtime behavior" note), but I'm not opening a PR with it given the autonomous-agent policy. Happy to have a maintainer or the reporter pick this up — posting the approach in case it saves investigation time.
